### PR TITLE
Add support for TI's LP_MSPM0G3507

### PR DESCRIPTION
**Pull request by msp-ti**
This contribution adds support for TI's MSPM0 devices starting with support for board LP_MSPM0G3507.
This PR currently includes support for:
* GPIO
* UART
* ~I2C~
* ~ADC~
* ~CAN~

### CODE REVIEW

**Review by henrikbrixandersen — DISMISSED** [review]
Hi, thank you for this contribution! However, please leave out I2C, ADC, and CAN support from the initial PR in order to reduce the number of code lines to review. 

Once the basic support has been reviewed and accepted, you can follow-up with individual PRs for adding more drivers. 

Keep in mind that these areas may need to be reviewed by different maintainers. Mixing it all together in one big PR makes it much harder to get it reviewed and approved. 

Please see https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html for further tips.

**Comment by msp-ti** [comment]
> Hi, thank you for this contribution! However, please leave out I2C, ADC, and CAN support from the initial PR in order to reduce the number of code lines to review.
> 
> Once the basic support has been reviewed and accepted, you can follow-up with individual PRs for adding more drivers.
> 
> Keep in mind that these areas may need to be reviewed by different maintainers. Mixing it all together in one big PR makes it much harder to get it reviewed and approved.
> 
> Please see https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html for further tips.

Sounds good. I resubmitted the PR after removing I2C, ADC and CAN, and fixed some copyright issues

**Review by gmarull — CHANGES_REQUESTED** [review]
please split the content in scoped commits, eg serial bindings, serial driver, soc dts files, soc, board, etc.

**Comment by fabiobaltieri** [comment]
@msp-ti for the missing hal files that are causing the CI failure, you can follow the process https://docs.zephyrproject.org/latest/develop/modules.html#process-for-submitting-changes-to-existing-modules to include the in-flight PR in this one, basically you have to change the ref in `west.yml` for `hal_ti` to `pull/51/head` (the PR number of https://github.com/zephyrproject-rtos/hal_ti/pull/51). You can do that as its own commit, once this one pulls and passes we can then merge the HAL PR and you can update this one with the final sha.

**Review by nordicjm — CHANGES_REQUESTED** [review]
Needs to be migrated to hwmv2

File: boards/arm/lp_mspm0g3507/support/MSPM0G3507.ccxml
Comment (vaishnavachath) [inline boards/arm/lp_mspm0g3507/support/MSPM0G3507.ccxml:19]: Doesn't CCS versions that support MSPM0G3507 come with default CCXML? is there a need to provide this here?

**Review by vaishnavachath — CHANGES_REQUESTED** [review]

**Comment by ddavidebor** [comment]
Just found out about this new line of Micros from digikey, very exciting ! Looking forward to testing them !

File: boards/arm/lp_mspm0g3507/support/MSPM0G3507.ccxml
Comment (msp-ti) [reply boards/arm/lp_mspm0g3507/support/MSPM0G3507.ccxml:19]: I don't believe this is necessary. I'll go ahead and delete it.

File: soc/ti/mspm0/mspm0g3xxx/soc.c
Comment (fabiobaltieri) [inline soc/ti/mspm0/mspm0g3xxx/soc.c:14]: do you need these commends? the first line does not add anything, just drop it, the other one up to you but I'd drop it as well, the function name seems descriptive enough

**Comment by msp-ti** [comment]
Fixed the problems with style-checks. Still other work to be done

File: boards/ti/lp_mspm0g3507/board.cmake
Comment (msp-ti) [reply boards/ti/lp_mspm0g3507/board.cmake:6]: We can leave out openOCD support for now if that's easier? It doesn't have public support so we're working on adding it at the moment

File: boards/ti/lp_mspm0g3507/lp_mspm0g3507_defconfig
Comment (msp-ti) [reply boards/ti/lp_mspm0g3507/lp_mspm0g3507_defconfig]: Makes sense. Made that change

File: boards/ti/lp_mspm0g3507/board.cmake
Comment (fabiobaltieri) [reply boards/ti/lp_mspm0g3507/board.cmake:6]: Up to you but I think it's fine to keep it, just that I would group the script in a config file

File: boards/ti/lp_mspm0g3507/board.yml
Comment (msp-ti) [reply boards/ti/lp_mspm0g3507/board.yml:5]: This is the specific board we offer though, I'm not sure if they would want to choose other boards, do we put more socs here?

Comment (fabiobaltieri) [reply boards/ti/lp_mspm0g3507/board.yml:5]: `name:` on the top (line 2) is the board one and that should be specific, the one here (line 5) is the soc and you want to have wildcard for variation that are expressed in devicetree (typically that's flash, ram, peripherals...)

File: soc/ti/mspm0/mspm0g3xxx/Kconfig
Comment (nordicjm) [inline soc/ti/mspm0/mspm0g3xxx/Kconfig:16]: help text for this goes in `Kconfig.soc`

File: soc/ti/mspm0/mspm0g3xxx/Kconfig.defconfig.mspm0g3507
Comment (nordicjm) [inline soc/ti/mspm0/mspm0g3xxx/Kconfig.defconfig.mspm0g3507:9]: `Kconfig.soc`

File: soc/ti/mspm0/mspm0g3xxx/Kconfig.soc
Comment (nordicjm) [inline soc/ti/mspm0/mspm0g3xxx/Kconfig.soc:10]: help text indent is 1x tab followed by 2x spaces

File: soc/ti/mspm0/mspm0g3xxx/linker.ld
Comment (nordicjm) [inline soc/ti/mspm0/mspm0g3xxx/linker.ld:8]: just link the file correctly in the cmake and delete this file

File: dts/bindings/pinctrl/ti,mspm0-pinctrl.yaml
Comment (nordicjm) [inline dts/bindings/pinctrl/ti,mspm0-pinctrl.yaml:120]: don't have "enable" as part of the names, it's implicit by the option itself

File: drivers/gpio/gpio_mspm0.c
Comment (nordicjm) [inline drivers/gpio/gpio_mspm0.c:81]:
```suggestion
#error "Series lookup table not supported"
```

File: drivers/gpio/gpio_mspm0.c
Comment (nordicjm) [inline drivers/gpio/gpio_mspm0.c:18]:
```suggestion
```

File: boards/ti/lp_mspm0g3507/docs/img/lp_mspm0g3507.png
Comment (nordicjm) [inline boards/ti/lp_mspm0g3507/docs/img/lp_mspm0g3507.png:1]: convert to webp and put it through https://tinypng.com/

File: boards/ti/lp_mspm0g3507/lp_mspm0g3507_defconfig
Comment (nordicjm) [inline boards/ti/lp_mspm0g3507/lp_mspm0g3507_defconfig:16]:
```suggestion
# Enable gpio driver
CONFIG_GPIO=y

# Enable UART driver
CONFIG_SERIAL=y

# Enable console
CONFIG_CONSOLE=y
CONFIG_UART_CONSOLE=y

# Enable pinctrl
CONFIG_PINCTRL=y
```

File: boards/ti/lp_mspm0g3507/lp_mspm0g3507.yaml
Comment (nordicjm) [inline boards/ti/lp_mspm0g3507/lp_mspm0g3507.yaml:1]:
```suggestion
identifier: lp_mspm0g3507/mspm0g3507
```

File: boards/ti/lp_mspm0g3507/lp_mspm0g3507.yaml
Comment (nordicjm) [inline boards/ti/lp_mspm0g3507/lp_mspm0g3507.yaml:3]: for the .dts, .yaml and _defconfig file please rename to `lp_mspm0g3507_mspm0g3507` then .dts, .yaml or _defconfig

**Review by nordicjm — CHANGES_REQUESTED** [review]

File: soc/ti/mspm0/mspm0g3xxx/soc.c
Comment (JFMSP) [reply soc/ti/mspm0/mspm0g3xxx/soc.c]: I'll clean up this file actually

File: soc/ti/mspm0/mspm0g3xxx/soc.c
Comment (JFMSP) [reply soc/ti/mspm0/mspm0g3xxx/soc.c]: This file can get cleaned up from this state and be a little more readable. It was mimicking one of our configuration tools which is outside of zephyr

File: dts/bindings/pinctrl/ti,mspm0-pinctrl.yaml
Comment (JFMSP) [reply dts/bindings/pinctrl/ti,mspm0-pinctrl.yaml:120]: input-enable has enable?

File: boards/ti/lp_mspm0g3507/lp_mspm0g3507.yaml
Comment (JFMSP) [reply boards/ti/lp_mspm0g3507/lp_mspm0g3507.yaml:1]: I don't fully understand this change for the identifier. Why do I want to do this?

File: boards/ti/lp_mspm0g3507/lp_mspm0g3507.yaml
Comment (JFMSP) [reply boards/ti/lp_mspm0g3507/lp_mspm0g3507.yaml:3]: I don't understand the rationale behind this rename.

File: boards/ti/lp_mspm0g3507/lp_mspm0g3507.yaml
Comment (nordicjm) [reply boards/ti/lp_mspm0g3507/lp_mspm0g3507.yaml:1]: Do you mean one specific soc as in `mspm0g3507` or do you mean one specific soc family as in `mspm0g350x` where x can be multiple things? The product page for this board isn't great because it lists a whole bunch of devices and doesn't specify if they come on the board of if they're just compatible. If this only supports `mspm0g3507` then this request can be ignored and marked as resolved, if it supports multiple then it must be added

File: boards/ti/lp_mspm0g3507/lp_mspm0g3507.yaml
Comment (nordicjm) [reply boards/ti/lp_mspm0g3507/lp_mspm0g3507.yaml:3]: As comment above

File: boards/ti/lp_mspm0g3507/board.cmake
Comment (vaishnavachath) [reply boards/ti/lp_mspm0g3507/board.cmake:6]: @msp-ti if OpenOCD support is not public, what would be the recommendation for someone trying to flash to the board? would `west flash` work ? also since these pre load commands are not Zephyr specific can that be part of your generic OpenOCD support?

File: boards/ti/lp_mspm0g3507/docs/index.rst
Comment (vaishnavachath) [inline boards/ti/lp_mspm0g3507/docs/index.rst:89]: No `west flash` support? I believe having a way  to build and flash examples to a platform without relying on out of tree vendor tools is a basic requirement for a new platform being added. Ideally I will recommend to add OpenOCD support, but if that is not possible, is it possible to to invoke uniflash CLI from board.cmake?)

File: boards/ti/lp_mspm0g3507/docs/img/lp_mspm0g3507.png
Comment (kartben) [inline boards/ti/lp_mspm0g3507/docs/img/lp_mspm0g3507.png:1]: consider using webp format to keep file size minimal - see ex. https://developers.google.com/speed/webp/download
file size can easily be halved

```
 ~#@❯  cwebp lp_mspm0g3507.png -o out.png                                                                                                                  ❮  
Saving file 'out.png'
File:      lp_mspm0g3507.png
Dimension: 500 x 281 (with alpha)
Output:    46592 bytes Y-U-V-All-PSNR 36.43 37.20 35.05   36.28 dB
           (2.65 bpp)
```

File: temp.stylefix
Comment (kartben) [inline temp.stylefix:1]: this would need to go away :)

Comment (kartben) [inline temp.stylefix:1]: this file would need to go away :)

**Review by kartben — CHANGES_REQUESTED** [review]

**Comment by kartben** [comment]
@msp-ti are you planning to come back to this PR?

**Comment by msp-ti** [comment]
> @msp-ti are you planning to come back to this PR?

Yes the plan is to pick this back up soon. We were working to add openOCD and had some other items come up that required work, but the hope is to get this PR back merged. I apologise for the delay

**Comment by Ayush1325** [comment]
@msp-ti Any updates?

File: temp.stylefix
Comment (msp-ti) [reply temp.stylefix:1]: You're right. Thanks for pointing that out

**Comment by msp-ti** [comment]
Sorry for the delay, I am working to port this PR to the newer updates with Zephyr and add some additional information

**Comment by msp-ti** [comment]
Reopening

File: boards/ti/lp_mspm0g3507/docs/index.rst
Comment (msp-ti) [reply boards/ti/lp_mspm0g3507/docs/index.rst:89]: Added comments to mention west flash. Open OCD support is there but pending a merge to be default

File: boards/ti/lp_mspm0g3507/board.cmake
Comment (msp-ti) [reply boards/ti/lp_mspm0g3507/board.cmake:6]: West Flash board can be achieved. Made note in the description

**Pull request closed without merging**
